libbpf-rs/src/perf_buffer.rs

```diff
@@ -7,44 +7,39 @@ use crate::*;
 pub struct PerfBuffer {}
 
 impl PerfBuffer {
-    pub fn new(_map: Map, _pages: usize, _opts: PerfBufferOpts) -> Self {
-        unimplemented!();
-    }
-
-    pub fn poll(&mut self, _timeout: Duration) {
-        unimplemented!();
-    }
-}
-
-/// Options and callbacks to configure [`PerfBuffer`].
-///
-/// Some callbacks provide raw byte slices. You may find libraries such as
-/// [`plain`](https://crates.io/crates/plain) helpful.
-#[derive(Default)]
-pub struct PerfBufferOpts {}
-
-impl PerfBufferOpts {
-    pub fn new() -> Self {
+    pub fn new(_map: Map) -> Self {
         unimplemented!();
     }
 
     /// Callback to run when a sample is received.
     ///
-    /// Callback arguments are: (ctx, cpu, data).
-    pub fn sample_cb<F>(self, _cb: F) -> Self
+    /// This callback provides a raw byte slice. You may find libraries such as
+    /// [`plain`](https://crates.io/crates/plain) helpful.
+    ///
+    /// Callback arguments are: (cpu, data).
+    pub fn sample_cb<F>(&mut self, _cb: F) -> &mut Self
     where
-        F: FnMut(usize, u32, &[u8]),
+        F: FnMut(u32, &[u8]),
     {
         unimplemented!();
     }
 
     /// Callback to run when a sample is received.
     ///
-    /// Callback arguments are: (ctx, cpu, lost_count).
-    pub fn lost_cb<F>(self, _cb: F) -> Self
+    /// Callback arguments are: (cpu, lost_count).
+    pub fn lost_cb<F>(&mut self, _cb: F) -> &mut Self
     where
-        F: FnMut(usize, u32, u64),
+        F: FnMut(u32, u64),
     {
         unimplemented!();
     }
+
+    /// The number of pages to size the ring buffer.
+    pub fn pages(&mut self, _pages: usize) -> &mut Self {
+        unimplemented!();
+    }
+
+    pub fn poll(&mut self, _timeout: Duration) {
```

Comment:
can fail, also doesn't need mutable reference, probably

Reply:
If callbacks receive a mutable ref to PerfBuffer, I think it makes sense to keep this as mutable.

Reply:
do callbacks need mutable reference? what for?

Reply:
Do we need a reference back to the `PerfBuffer` at all? We don't currently have any properties that can really be changed after construction.

If we wanna add it later we can easily add a `poll_mut()` or `poll_ref()` method. But TBH I can't see a use case.

Reply:
yeah, I don't have specific use case to have any kind of references, but it's usually a good idea to have reference to an "object" that's calling a callback function. E.g., just in case someone has multiple PerfBuffers, but the same callback. Having reference to PerfBuffer would allow to differentiate between them, if necessary. 

But I think we are talking about different things here. PerfBuffer::poll needs &self -- I don't think PerfBuffer needs to be modified as a result of poll operation. What I was describing above is what should be passed to user-provided callback. There, I think, having PerfBuffer reference is useful.

Reply:
Sounds reasonable